### PR TITLE
Add signal-handling to allow proper log-rotation.

### LINKED ISSUE #420: Rotating verdaccio logs
This is the same question I asked for rlidwka/sinopia#472 -- I'm configuring Verdaccio to log into a file, which will need to be rotated.

Is there some way to signal to the program to reopen the log, or do I have to fully restart it? Thank you!

### PR DESCRIPTION
**Pull request by UnitedMarsupials**
**Type:** bug

The following has been addressed in the PR:

*  No mechanism exists to rotate Verdaccio logs without a full restart

**Description:**

Resolves #420

### CODE REVIEW

File: src/lib/logger.js [hunk 1]
```diff
@@ -93,6 +93,10 @@ function setup(logs) {
 		},
 	});
 
+       process.on('SIGUSR2', function () {
+               log.reopenFileStreams();
```
Comment (juanpicado) [inline src/lib/logger.js]: Check the linting is failing, `log` is not defined.

you can run locally `npm run lint` to double check.

Comment (UnitedMarsupials) [reply src/lib/logger.js]: Fixed. Lint passes now. Tested on Ubuntu Trusty.

**Comment by juanpicado** [comment]
Sorry, in order to release I need you update (https://github.com/verdaccio/verdaccio/blob/master/CONTRIBUTING.md#git-commit-guidelines) commit messages. We have an automated process, otherwise, the CI won't do the release.

**Comment by UnitedMarsupials** [comment]
> I need you update

I added the line to the best of my understanding. Feel free to correct it, if I misread the requirement(s).

**Comment by UnitedMarsupials** [comment]
Or should I have started a new entry -- for 2.6.7? Or 2.7.0?

**Comment by UnitedMarsupials** [comment]
Ok, I think, I got it... You don't want commits directly to `CHANGELOG.md` -- you want the commit-message(s) document what they are fixing/changing so the file can be automatically generated.

How about now?

**Comment by UnitedMarsupials** [comment]
I do not understand, why the test has failed now. The PR is back to #10a49e439ce1a5a5279f2223247cf47d6d1e85d5 , when the tests passed...

**Comment by juanpicado** [comment]
@UnitedMarsupials about the test. All test passed, do not worry about it.
About the commits( sorry I was in typing on my phone). We follow this convention https://conventionalcommits.org/ for commit messages, then, we don't have to update the changelog manually and thus CI also take advantage of such thing. So I'd ask again to amend your commit messages otherwise won't appear in the changelog.

**Comment by UnitedMarsupials** [comment]
> So, I'd ask again to amend your commit messages

Can one even do such amending for commits that have been pushed already? And is that necessary, since I've done subsequent commits with a message suitably formatted (I hope)? I'm talking about the https://github.com/verdaccio/verdaccio/pull/424/commits/baa4763a248053347c7648c067ee883eae37a61d

**Comment by juanpicado** [comment]
Yes, you can. I suggest read this guide https://help.github.com/articles/changing-a-commit-message/

in your case would be (then follow the guide)
```
git rebase -i HEAD~7
```
 then you end doing `git push --force` and the PR will be updated automatically 👍

**Comment by UnitedMarsupials** [comment]
> git rebase -i HEAD~7

I did that and now I can not push into my own branch any more. It is getting ridiculous.

Seriously, the entire fix was three lines of code and one line of documentation. You seem to agree, the change is good and useful -- please, _do whatever is needed_ to merge it in.

You do not need to give me credit for it, if that's holding you up. Thank you.

**Comment by juanpicado** [comment]
@UnitedMarsupials thanks for the PR. It's not only about credit, I'm just trying to keep the things in order around here, maintenance and help others to find quick answers. I'll merge since this is important to you and is a quite good contribution.

**Comment by UnitedMarsupials** [comment]
I hate to say, but there is a problem. When the signal arrives, the process crashes:
```
Logger.reopenFileStreams is not a function
```
I didn't notice this, because `upstart` on Ubuntu would immediately restart the service -- and I was only checking, whether the _currently running_ verdaccio has the correct file opened.

I'll have a new PR soon :-/

**Comment by juanpicado** [comment]
Cool, I'll be ready to release it :)

**Comment by UnitedMarsupials** [comment]
The problem is, bunyan can only reopen the regular file stream -- so the fancy logic of object-printing based on the setting of format to "`pretty`" will break.

Either we'd have to forgo the "pretty" functionality, when logging to a _file_ (fine with me), or I have to implement our own reopener...

**Comment by UnitedMarsupials** [comment]
Ok, I have the first variant ready -- forgo the pretty functionality and let bunyan handle file reopening the way it wants to:
```patch
index 5b2b38f..8dea413 100644
--- a/src/lib/logger.js
+++ b/src/lib/logger.js
@@ -39,6 +56,9 @@ function setup(logs) {
 		const stream = new Stream();
 		stream.writable = true;
 
+		if (target.level === undefined || target.level === 'http')
+			target.level = 35;
+
 		if (target.type === 'stdout' || target.type === 'stderr') {
 			// destination stream
 			const dest = target.type === 'stdout' ? process.stdout : process.stderr;
@@ -58,31 +78,24 @@ function setup(logs) {
 					dest.write(JSON.stringify(obj, Logger.safeCycles()) + '\n');
 				};
 			}
+			streams.push({
+				type: 'raw',
+				level: target.level,
+				stream: stream,
+			});
 		} else if (target.type === 'file') {
-			const dest = require('fs').createWriteStream(target.path, {flags: 'a', encoding: 'utf8'});
-			dest.on('error', function(err) {
-				Logger.emit('error', err);
+			streams.push({
+				type: 'file',
+				level: target.level,
+				path: target.path,
 			});
-			stream.write = function(obj) {
-				if (target.format === 'pretty') {
-					dest.write(print(obj.level, obj.msg, obj, false) + '\n');
-				} else {
-					dest.write(JSON.stringify(obj, Logger.safeCycles()) + '\n');
-				}
-			};
 		} else {
 			throw Error('wrong target type for a log');
 		}
 
-		if (target.level === 'http') target.level = 35;
-		streams.push({
-			type: 'raw',
-			level: target.level || 35,
-			stream: stream,
-		});
 	});
 
-	// buyan default configuration
+	// bunyan default configuration
 	const logger = new Logger({
 		name: pkgJSON.name,
 		streams: streams,
@@ -94,7 +107,7 @@ function setup(logs) {
 	});
 
        process.on('SIGUSR2', function() {
-               Logger.reopenFileStreams();
+               module.exports.logger.reopenFileStreams();
        });
 
 	module.exports.logger = logger;
```

This _should_ work, but does not because of googleapis/nodejs-logging/issues/14 (originally known as googleapis/nodejs-logging-bunyan#2), presumably fixed in googleapis/nodejs-logging/issues/15 and googleapis/nodejs-logging/issues/16.

Unfortunately, this is my first attempt at programming NodeJS and I do not know, how to use npm to get the newer bunyan. Need help, @juanpicado ...

**Comment by UnitedMarsupials** [comment]
Ok, I understand, what's happening now. First of all, you are calling the logger `setup()` twice -- before and after parsing the config-file. This will create a _two_ signal-handlers, unless a check is added.

More importantly, bunyan's code for creating children -- which you invoke in five different places to provide different colors for different subsystems -- creates multiple _copies_ of the opened streams. Reopening the parent's streams currently has no effect on these children -- _their_ copies of streams become invalid (node-bunyan/issues#563).

The _quick_ fix for lack of log-rotation is then thus:

1. Stop offering "pretty" log-messages for when the log-destination is a file
1. Stop using "subsystems" and refer to the one-and-only logger-object instead:
```diff
-    this.logger = Logger.logger.child({sub: 'auth'});
+    this.logger = Logger.logger;
```
Is this drop of pretty colors a suitable sacrifice? In my opinion, such coloring is utterly useless -- but I work with production servers and log-files -- not debugging the application from command-line.

If it is not acceptable, then, to avoid the first "sacrifice" we'd need to rescan the stream list readding our own `write()` methods to each file-stream after it gets reopened. Doable, but messy.

For the second sacrifice, we'd need to wait for bunyan to fix the children-creation -- I wish, I knew NodeJS well enough to do it myself...

**Comment by juanpicado** [comment]
Thanks @UnitedMarsupials for the research. Let me process all of this and I'll give you my feedback 👍  later on.

**Comment by juanpicado** [comment]
@UnitedMarsupials ok I tried to get into `bunyan` such I'm not an expert. 

First, colors, we need them. No way to remove colors. Are useless in some environments, but, quite helpful in local environments. We need a balance here.

I'm mostly thinking to support option number 2, at the end, "subsystems" are not useful IMO, but I wonder whether https://github.com/trentm/node-bunyan/pull/442 will be merged since has 1 year in the queue. 

I wonder why both sacrifices need to stop to use colors. 

Just a reminder, I appreciate the effort to you finding a fix of this, but it will be used only on 2.x. We have running https://github.com/verdaccio/verdaccio/pull/408 for the next big release.

**Pull request merged**